server/platform/services/cache/redis.go

```diff
@@ -0,0 +1,285 @@
+		r.client.B().Del().
+			Key(keys...).
+			Build(),
+	).Error()
+}
+
+func (r *Redis) Set(key string, value any) error {
+	return r.SetWithExpiry(key, value, 0)
+}
+
+// SetWithDefaultExpiry adds the given key and value to the store with the default expiry. If
+// the key already exists, it will overwrite the previous value
+func (r *Redis) SetWithDefaultExpiry(key string, value any) error {
+	return r.SetWithExpiry(key, value, r.defaultExpiry)
+}
+
+// SetWithExpiry adds the given key and value to the cache with the given expiry. If the key
+// already exists, it will overwrite the previous value
+func (r *Redis) SetWithExpiry(key string, value any, ttl time.Duration) error {
+	now := time.Now()
+	defer func() {
+		elapsed := float64(time.Since(now)) / float64(time.Second)
+		if r.metrics != nil {
+			r.metrics.ObserveRedisEndpointDuration(r.name, "Set", elapsed)
+		}
+	}()
+	var buf []byte
+	var err error
+	// We use a fast path for hot structs.
+	if msgpVal, ok := value.(msgp.Marshaler); ok {
+		buf, err = msgpVal.MarshalMsg(nil)
+	} else {
+		// Slow path for other structs.
+		buf, err = msgpack.Marshal(value)
+	}
+	if err != nil {
+		return err
+	}
+
+	return r.client.Do(context.Background(),
+		r.client.B().Set().
+			Key(r.name+":"+key).
+			Value(rueidis.BinaryString(buf)).
+			Ex(ttl).
+			Build(),
+	).Error()
+}
+
+// Get the content stored in the cache for the given key, and decode it into the value interface.
+// Return ErrKeyNotFound if the key is missing from the cache
+func (r *Redis) Get(key string, value any) error {
+	now := time.Now()
+	defer func() {
+		elapsed := float64(time.Since(now)) / float64(time.Second)
+		if r.metrics != nil {
+			r.metrics.ObserveRedisEndpointDuration(r.name, "Get", elapsed)
+		}
+	}()
+	val, err := r.client.DoCache(context.Background(),
+		r.client.B().Get().
+			Key(r.name+":"+key).
+			Cache(),
+		5*time.Minute,
+	).AsBytes()
+	if err != nil {
+		if rueidis.IsRedisNil(err) {
+			return ErrKeyNotFound
+		}
+		return err
+	}
+
+	// We use a fast path for hot structs.
+	if msgpVal, ok := value.(msgp.Unmarshaler); ok {
+		_, err := msgpVal.UnmarshalMsg(val)
+		return err
+	}
+
+	// This is ugly and makes the cache package aware of the model package.
+	// But this is due to 2 things.
+	// 1. The msgp package works on methods on structs rather than functions.
+	// 2. Our cache interface passes pointers to empty pointers, and not pointers
+	// to values. This is mainly how all our model structs are passed around.
+	// It might be technically possible to use values _just_ for hot structs
+	// like these and then return a pointer while returning from the cache function,
+	// but it will make the codebase inconsistent, and has some edge-cases to take care of.
+	switch v := value.(type) {
+	case **model.User:
+		var u model.User
+		_, err := u.UnmarshalMsg(val)
+		*v = &u
+		return err
+	case *map[string]*model.User:
+		var u model.UserMap
+		_, err := u.UnmarshalMsg(val)
+		*v = u
+		return err
+	}
```

Comment:
I don't really understand why we need this only for these two types. For my own understanding: could you expand a bit on this?

Reply:
All of this started when I added `msgp` which would auto-generate fast marshal/unmarshal code which didn't use reflect. Now usually, you wouldn't need this and the code from L117-120 would just do the job. But the issue is that the interface is only implemented for `*model.User`, not `**model.User`, which brings us to the comment I mentioned above:

> // 2. Our cache interface passes pointers to empty pointers, and not pointers
	// to values. This is mainly how all our model structs are passed around.
	// It might be technically possible to use values _just_ for hot structs
	// like these and then return a pointer while returning from the cache function,
	// but it will make the codebase inconsistent, and has some edge-cases to take care of.

So that's why we do this. Can we fix this? Sure. But then we need to change a _lot_ of other things across the entire codebase as well.

Reply:
> But the issue is that the interface is only implemented for `*model.User`, not `**model.User`

Ah! That's the thing I was missing. That makes sense. Again, just for the sake of satisfying my curiosity: can we generate another interface for `**model.User`?

Reply:
That's an interesting proposition. So the thing is that these methods are auto-generated, I don't want to change that. But I will take a look to see if I can wrap the `*model.User` method with `**model.User`.